src/Components/Components/src/ComponentsActivitySource.cs

```diff
@@ -0,0 +1,177 @@
+        if (parentActivity is not null && parentActivity.OperationName == "Microsoft.AspNetCore.Hosting.HttpRequestIn" && parentActivity.Recorded)
+        {
+            return parentActivity.Context;
+        }
+        return default;
+    }
+
+    public Activity? StartCircuitActivity(string circuitId, ActivityContext httpContext)
+    {
+        _circuitId = circuitId;
+
+        var activity = ActivitySource.CreateActivity(OnCircuitName, ActivityKind.Internal, parentId: null, null, null);
+        if (activity is not null)
+        {
+            if (activity.IsAllDataRequested)
+            {
+                if (_circuitId != null)
+                {
+                    activity.SetTag("aspnetcore.components.circuit.id", _circuitId);
+                }
+                if (httpContext != default)
+                {
+                    activity.AddLink(new ActivityLink(httpContext));
+                }
+            }
+            activity.DisplayName = $"Circuit {circuitId ?? ""}";
+            activity.Start();
+            _circuitContext = activity.Context;
+        }
+        return activity;
+    }
+
+    public void FailCircuitActivity(Activity? activity, Exception ex)
+    {
+        _circuitContext = default;
+        if (activity != null && !activity.IsStopped)
+        {
+            activity.SetTag("error.type", ex.GetType().FullName);
+            activity.SetStatus(ActivityStatusCode.Error);
+            activity.Stop();
+        }
+    }
+
+    public Activity? StartRouteActivity(string componentType, string route)
+    {
+        if (_httpContext == default)
+        {
+            _httpContext = CaptureHttpContext();
+        }
+
+        var activity = ActivitySource.CreateActivity(OnRouteName, ActivityKind.Internal, parentId: null, null, null);
+        if (activity is not null)
+        {
+            if (activity.IsAllDataRequested)
+            {
+                if (_circuitId != null)
+                {
+                    activity.SetTag("aspnetcore.components.circuit.id", _circuitId);
+                }
+                if (componentType != null)
+                {
+                    activity.SetTag("aspnetcore.components.type", componentType);
+                }
+                if (route != null)
+                {
+                    activity.SetTag("aspnetcore.components.route", route);
+                }
+                if (_httpContext != default)
+                {
+                    activity.AddLink(new ActivityLink(_httpContext));
+                }
+                if (_circuitContext != default)
+                {
+                    activity.AddLink(new ActivityLink(_circuitContext));
+                }
+            }
+
+            activity.DisplayName = $"Route {route ?? "[unknown path]"} -> {componentType ?? "[unknown component]"}";
+            activity.Start();
+            _routeContext = activity.Context;
+        }
+        return activity;
+    }
```

Comment:
This should go on the Server assembly.

Reply:
If we move it to different class/instance, we would not be able to use `_circuitId` and `_circuitContext`. And we would have to create lot of semi-public interfaces to pass that info between those classes. Are you sure it's worth it ?

Reply:
Isn't the CircuitId on a parent activity? (like start circuit or something like that)

All the `_httpContext`, `_circuitContext` on that event is at the wrong level of abstraction. I don't know if it can be set on the parent activity and flowed into this one, but the route activity should not know about _circuitContext or _httpContext.

To put it in a different way, this method/class should not be responsible for resolving the parent activity to link to. That should be provided somehow. We might need public API for that.

We don't want to keep adding to this list when/if we add new render modes. For example, I imagine WebView migth also want to track its own parent activity and it would require us to add it here.

Reply:
Let's address this in next PR.

I agree that we should not break existing expectations about layers. On the other hand, I would like to keep this as simple as possible. This is internal class and I don't prefer lofty layered architectures for simple and perf sensitive classes like this one.